core/cache/model.go

```diff
@@ -121,6 +137,27 @@ func (m *Model) removeApplication(ch RemoveApplication) {
 	m.mu.Unlock()
 }
 
+// updateCharm adds or updates the charm in the model.
+func (m *Model) updateCharm(ch CharmChange) {
+	m.mu.Lock()
+
+	charm, found := m.charms[ch.CharmURL]
+	if !found {
+		charm = newCharm(m.metrics, m.hub)
+		m.charms[ch.CharmURL] = charm
+	}
+	charm.setDetails(ch)
+
+	m.mu.Unlock()
```

Comment:
Why not `defer m.mu.Unlock()`?

Reply:
Defer has an overhead, so it's better to use it only when you don't know where the method might exit.

Reply:
citation needed, not saying I don't agree, but is this not premature optimisation for the sake of a defer?

Reply:
https://github.com/golang/go/issues/14939